File: apollo-runtime/src/main/java/com/apollographql/android/cache/http/ResponseBodyCacheSink.java
package com.apollographql.android.cache.http;

import java.io.IOException;

import okio.Buffer;
import okio.BufferedSink;
import okio.ForwardingSink;

class ResponseBodyCacheSink extends ForwardingSink {
  private boolean failed;

  ResponseBodyCacheSink(BufferedSink delegate) {
    super(delegate);
  }

  @Override public void write(Buffer source, long byteCount) throws IOException {
    if (failed) return;
    try {
      super.write(source, byteCount);
    } catch (Exception e) {
      failed = true;
      onException(e);
    }
  }

  @Override public void flush() throws IOException {
    if (failed) return;
    try {
      super.flush();
    } catch (Exception e) {
      failed = true;
      onException(e);
    }
  }

  @Override public void close() throws IOException {
    if (failed) return;
    try {
      super.close();
    } catch (Exception e) {
      failed = true;
      onException(e);
    }
  }

  void copyFrom(Buffer buffer, long offset, long bytesCount) {
    if (failed) return;
    try {
      BufferedSink outSink = (BufferedSink) delegate();
      buffer.copyTo(outSink.buffer(), offset, bytesCount);
      outSink.emitCompleteSegments();
    } catch (Exception e) {
      failed = true;
      onException(e);
    }
  }

  void onException(Exception e) {

  }
}

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/ResponseBodyCacheSink.kt
package com.apollographql.apollo3.cache.http

import okio.Buffer
import okio.BufferedSink
import okio.ForwardingSink
import java.io.IOException

internal abstract class ResponseBodyCacheSink(delegate: BufferedSink?) : ForwardingSink(delegate!!) {
  private var failed = false
  @Throws(IOException::class)
  override fun write(source: Buffer, byteCount: Long) {
    if (failed) return
    try {
      super.write(source, byteCount)
    } catch (e: Exception) {
      failed = true
      onException(e)
    }
  }

  @Throws(IOException::class)
  override fun flush() {
    if (failed) return
    try {
      super.flush()
    } catch (e: Exception) {
      failed = true
      onException(e)
    }
  }

  @Throws(IOException::class)
  override fun close() {
    if (failed) return
    try {
      super.close()
    } catch (e: Exception) {
      failed = true
      onException(e)
    }
  }

  fun copyFrom(buffer: Buffer, offset: Long, bytesCount: Long) {
    if (failed) return
    try {
      val outSink = delegate as BufferedSink
      buffer.copyTo(outSink.buffer(), offset, bytesCount)
      outSink.emitCompleteSegments()
    } catch (e: Exception) {
      failed = true
      onException(e)
    }
  }

  abstract fun onException(e: Exception?)
}